privacy/i2p.md

```diff
@@ -0,0 +1,317 @@
+        || (echo "Error starting TOR!", QUBES_IP="127.0.0.1")
+
+        #I have no idea what tinyproxy is used for.
+        #Killing it may be a bad idea.
+        systemctl stop tinyproxy
+        killall tinyproxy
+        #start tinyproxy with custom config-file
+        tinyproxy -c /rw/config/tinyproxy.conf
+
+        systemctl stop dnsmasq
+        killall dnsmasq
+        dnsmasq --address=/.i2p/$TARGET_IP --server 127.0.0.1#$TOR_DNS_PORT
+
+    else
+      echo "$QUBES_IP is localhost. Not starting!"
+    fi
+
+The dnsmasq commandline-options tell it to resolve any `.i2p` toplevel domain to `$TARGET_IP` and query TOR's DNS server for other domains.
+
+Relevant code of `/rw/config/qubes-firewall-user-script`:
+
+    !#/bin/bash
+
+    source /rw/config/config.sh
+    
+    #stop accidential forwards
+    iptables -F FOWARD
+    iptables -t nat -F
+
+    if [ "$QUBES_IP" == "127.0.0.1" ]; then
+      exit 1
+    fi
+
+    #allow local redirect
+    sysctl net.ipv4.ip_forward=1
+    sysctl net.ipv4.conf.all.route_localnet=1
+
+    #accept TOR, I2P and tinyproxy
+    iptables -I INPUT -i vif+ -p tcp -m multiport --dport $TOR_TRANS_PORT,$TOR_CTRL_PORT,$I2P_HTTP,$I2P_PORTS,$TINY_PORT -j ACCEPT
+    iptables -I INPUT -i vif+ -p udp -m multiport --dport 53,$TOR_DNS_PORT -j ACCEPT
+
+    #routing
+    #I2P IS FOR ME!
+    iptables -t nat -A PREROUTING -i vif+ -p tcp -d $TARGET_IP -m multiport --dport $I2P_HTTP,$I2P_PORTS -j REDIRECT
+
+    #HTTP is for tinyproxy!
+    iptables -t nat -A PREROUTING -i vif+ -p tcp -d $TARGET_IP --dport 80 -j REDIRECT --to-port $TINY_PORT
+
+    #ALL TCP IS FOR TOR!
+    iptables -t nat -A PREROUTING -i vif+ -p tcp -j REDIRECT --to-port $TOR_TRANS_PORT
+
+    #local DNS resolution
+    iptables -t nat -I PREROUTING -i vif+ -p udp --dport 53 -j REDIRECT --to-port 53
+
+
+# Usage #
+Pick any AppVM to access the I2P network or TOR hidden services. Set it's net-vm to the I2P proxyVM.
+Access the router-console by opening a browser and visiting `router.i2p:7567`.
+(Or, truly, any `.i2p`-domain at port :7657)  
+Check I2P is up and has enough connections.
+Click on one of the common pages on the routers home screen, e.g. i2pwiki.i2p!
+
+# Final thoughts #
+
+## Troubleshooting ##
+Generally, most Problems can be resolved by waiting long enough for I2P to gather enough connections or restarting I2P.
+
+#### 404 - tinyproxy can't reach upstream proxy ####
+I2P refuses connections.
+Either it doesn't have enough connections to answer comfortably (in which case, just wait!), or I2P stopped for some reason.
+In which case, open a shell in your proxyVM and type:
+
+    sudo systemctl start i2p
+
+#### I2P won't start! ####
+Check your qubes-service settings. If those are in order, check 
+
+## Personal notes ##
+
+Tor may be replaced with qubes-tor, I did not try that.
+
+Getting I2P to run in a whonix-gw would be great, but I have had no time to get into the whonix routing and didn't find a simple, comprehensive explanation.
```

Comment:
Btw Whonix community is working on that independently but I am not following the current status.

https://forums.whonix.org/t/i2p-integration